### PR TITLE
[CLIPPY] Fix the last portion of rules and enable CI check

### LINKED ISSUE #45: Clean the project using clippy
- fix issues suggested by https://github.com/rust-lang/rust-clippy
- setup linter pipeline executed on project test

### PR DESCRIPTION
**Pull request by kgrech**
This closes #45 

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [X] Does your submission pass tests?
2. [X] Have you lint your code locally using ``cargo fmt`` command prior to submission?

### CODE REVIEW

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -25,6 +25,8 @@ jobs:
       run: sudo apt-get install clang libopenblas-dev libgfortran-9-dev
     - name: Check code formatting
       run: cargo fmt --all -- --check
+    - name: Check cargo clippy warnings
+      run: cargo clippy --all-targets --all-features -- -D warnings
```
Comment (kgrech) [inline .github/workflows/rust.yml:29]: This compiles the whole project and sadly the results of this compilation are not using during the test. So maybe it worth moving ``fmt`` and ``clippy`` into the separate action workflow. Will probably do in the separate PR

File: lib/segment/src/index/struct_payload_index.rs [hunk 2]
```diff
@@ -217,14 +219,14 @@ impl StructPayloadIndex {
         Ok(field_indexes)
     }
 
-    fn build_and_save(&mut self, field: &PayloadKeyType) -> OperationResult<()> {
-        if !self.config.indexed_fields.contains(field) {
-            self.config.indexed_fields.push(field.clone());
+    fn build_and_save(&mut self, field: PayloadKeyTypeRef) -> OperationResult<()> {
+        if !self.config.indexed_fields.iter().any(|x| x == field) {
```
Comment (generall) [inline lib/segment/src/index/struct_payload_index.rs:223]: I think `contains` is more readable in this case. Can you pls refer the explanation for this change.

File: lib/segment/src/index/struct_payload_index.rs [hunk 3]
```diff
@@ -241,16 +243,16 @@ impl PayloadIndex for StructPayloadIndex {
         self.config.indexed_fields.clone()
     }
 
-    fn set_indexed(&mut self, field: &PayloadKeyType) -> OperationResult<()> {
-        if !self.config.indexed_fields.contains(field) {
-            self.config.indexed_fields.push(field.clone());
+    fn set_indexed(&mut self, field: PayloadKeyTypeRef) -> OperationResult<()> {
+        if !self.config.indexed_fields.iter().any(|x| x == field) {
```
Comment (generall) [inline lib/segment/src/index/struct_payload_index.rs:247]: same

**Review by generall — CHANGES_REQUESTED** [review]

File: lib/segment/src/index/struct_payload_index.rs (hunk 3 above)
Comment (kgrech) [reply lib/segment/src/index/struct_payload_index.rs:247]: Agree, but there is an unfortunate problem that contains method on Vec<String> accepts only &String and not &str. I did not find a way to convert &str to &String without creating the String,  which is a copying operation. Do you have any idea? Performance wise contains doing exactly the same as the above code .

Comment (generall) [reply lib/segment/src/index/struct_payload_index.rs:247]: Oh, I didn't realize it was actually a new type at first glance. In this case, readability is not so critical compared to consistency.

**Review by generall — APPROVED** [review]
